WebKit fails to render extreme border-radius

vitorroriz opened the pull request:
#### 8a3db1af14c197a24ab0425903d026408ff569f5
<pre>
WebKit fails to render extreme border-radius
<a href="https://bugs.webkit.org/show_bug.cgi?id=244638">https://bugs.webkit.org/show_bug.cgi?id=244638</a>
rdar://99668793

Reviewed by Antti Koivisto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable.html: Added.
Testing that child-background doesn&apos;t bleed through its parent border for a inner-border-radius that is larger than the content rect.

* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOneBorderSide):
Improving readability.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::clipContentForBorderRadius):
(WebCore::RenderBox::pushContentsClip):
The bug can be fixed by adjusting the inner-border radius, if not renderable, here. However, with we adjust it just here the bug can still be triggered in
other situations, for example, when applying a clip-path in an element with non-renderable border. Therefore, we are adjusting the non-renderable inner-border
directly in RenderStyle::getRoundedInnerBorderFor. Here, we are adding a helper function that still calls getRoundedInnerBorderFor for improving readability.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::getRoundedInnerBorderFor):
Adjusting the inner-border-radii for non-renderable inner-border.

Canonical link: <a href="https://commits.webkit.org/256943@main">https://commits.webkit.org/256943@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a21d79b0a3c8ed0cbefbf9890af5b3d8f3eb2fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6545 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30430 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106801 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167067 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101250 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6841 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35281 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89676 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103485 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102944 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83909 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32123 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86989 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75049 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/562 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/546 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21740 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4785 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5347 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1789 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41068 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/rendering/BorderPainter.h:
```diff
@@ -76,5 +76,6 @@ class BorderPainter {
 };
 
 LayoutRect shrinkRectByOneDevicePixel(const GraphicsContext&, const LayoutRect&, float devicePixelRatio);
+RoundedRect calculateAdjustedInnerBorder(const RoundedRect&innerBorder, BoxSide);
```
This could be static function in the BorderPainter scope.

vitorroriz replied:
It is being used re-used by RenderBox

anttijk commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2021,8 +2021,18 @@ bool RenderBox::pushContentsClip(PaintInfo& paintInfo, const LayoutPoint& accumu
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Bottom).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Left).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Right).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+        }
```
Maybe you have a helper for this that takes GraphicsContext a parameter, clipContentForBorderRadius or something.

smfr replied:
It seems wrong to compute a clip for each side (ending up with the intersection of 4 clips) rather than just having logic that fixes up the rounded rect.

smfr replied:
Why not just do:
```
diff --git a/Source/WebCore/rendering/RenderBox.cpp b/Source/WebCore/rendering/RenderBox.cpp
index 704c4286a85d4d304024cf7043d8f82637b39c34..6e6b63fc0229d372080fe08ae647ea8dfa778f89 100644
--- a/Source/WebCore/rendering/RenderBox.cpp
+++ b/Source/WebCore/rendering/RenderBox.cpp
@@ -2021,8 +2021,13 @@ bool RenderBox::pushContentsClip(PaintInfo& paintInfo, const LayoutPoint& accumu
     float deviceScaleFactor = document().deviceScaleFactor();
     FloatRect clipRect = snapRectToDevicePixels((isControlClip ? controlClipRect(accumulatedOffset) : overflowClipRect(accumulatedOffset, nullptr, IgnoreOverlayScrollbarSize, paintInfo.phase)), deviceScaleFactor);
     paintInfo.context().save();
-    if (style().hasBorderRadius())
-        paintInfo.context().clipRoundedRect(style().getRoundedInnerBorderFor(LayoutRect(accumulatedOffset, size())).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+    if (style().hasBorderRadius()) {
+        auto innerRoundedRect = style().getRoundedInnerBorderFor(LayoutRect(accumulatedOffset, size())).pixelSnappedRoundedRectForPainting(deviceScaleFactor);
+        if (!innerRoundedRect.isRenderable())
+            innerRoundedRect.adjustRadii();
+        paintInfo.context().clipRoundedRect(innerRoundedRect);
+    }
+    
     paintInfo.context().clip(clipRect);
 
     if (paintInfo.phase == PaintPhase::EventRegion)
```

vitorroriz replied:
Fixed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable.html:
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <link rel="match" href="inner-border-ref.html">
```
The match link here seems wrong, should be `inner-border-non-renderable-ref.html`

vitorroriz replied:
Fixed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable-ref.html:
```diff
@@ -0,0 +1,21 @@
+            border: 30px solid green;
+            border-top-color: gold;
+            border-top-right-radius: 150px 267px;
+            /* border-bottom-left-radius: 150px 267px; */
```
Is this commented out code useful? This whole file should probably match the -expected.html

vitorroriz replied:
Fixed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable.html:
```diff
@@ -0,0 +1,29 @@
+<html>
+<head>
+    <link rel="match" href="inner-border-ref.html">
+    <meta name=fuzzy content="0-144;0-484">
```
can you include maxDifference=/totalPixels= for clarity?

vitorroriz replied:
We don't need fuzziness anymore.

vitorroriz committed:
WebKit fails to render extreme border-radius
https://bugs.webkit.org/show_bug.cgi?id=244638
rdar://99668793

Reviewed by Antti Koivisto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/inner-border-non-renderable.html: Added.
Testing that child-background doesn't bleed through its parent border for a inner-border-radius that is larger than the content rect.

* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOneBorderSide):
Improving readability.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::clipContentForBorderRadius):
(WebCore::RenderBox::pushContentsClip):
The bug can be fixed by adjusting the inner-border radius, if not renderable, here. However, with we adjust it just here the bug can still be triggered in
other situations, for example, when applying a clip-path in an element with non-renderable border. Therefore, we are adjusting the non-renderable inner-border
directly in RenderStyle::getRoundedInnerBorderFor. Here, we are adding a helper function that still calls getRoundedInnerBorderFor for improving readability.

* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::getRoundedInnerBorderFor):
Adjusting the inner-border-radii for non-renderable inner-border.

Canonical link: https://commits.webkit.org/256943@main
